### ISSUE TITLE
Cannot exclude files from update by name or suffix in v 1.19 with goal license:update-file-header

### ISSUE THREAD

**Issue by chrisinmtown** [mojohaus/license-maven-plugin#311]
I would like to exclude JSON files from being decorated by l-m-p version 1.19 and mvn version 3.5.2 when run as 'mvn license:update-file-header'.  I started with a plugin configuration from the L-M-P example [here](https://www.mojohaus.org/license-maven-plugin/examples/update-file-header-config.html).

The identical problem was reported in #16 on an older version of LMP but closed without action.

I run the plugin like this, trying to use it as a one-and-done tool:

```
    mvn -DdryRun=true license:update-file-header
````

I set this up in a tiny project with exactly one file, namely angular.json.  The dry run output always shows it processing one file.  If I remove the dryRun option that file gets mangled.

I turned on verbose and saw this output, none of my added patterns are in there, if that matters.

```
[INFO] Will use default excludes[**/*.zargo, **/*.uml, **/*.umldi, **/*.xmi, **/*.img, **/*.png, **/*.jpg, **/*.jpeg, **/*.gif, **/*.zip, **/*.jar, **/*.war, **/*.ear, **/*.tgz, **/*.gz]
```

Please advise if there's an error between my keyboard and my chair :) thanks in advance.

```
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>license-maven-plugin</artifactId>
        <version>1.19</version>
        <configuration>
          <licenseName>gpl_v3</licenseName>
	  <organizationName>The Epoch</organizationName>
	  <inceptionYear>1970</inceptionYear>
          <verbose>false</verbose>
        </configuration>
        <executions>
          <execution>
            <id>first</id>
            <goals>
              <goal>update-file-header</goal>
            </goals>
            <phase>process-sources</phase>
            <configuration>
              <licenseName>gpl_v3</licenseName>
              <excludes>
                <exclude>angular.json</exclude>
                <exclude>*.json</exclude>
                <exclude>**/*.json</exclude>
              </excludes>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

**Comment by ppalaga** [comment]
Thank for the report @chrisinmtown.

We do have an integration test for `update-file-header` with an `<exclude>` https://github.com/mojohaus/license-maven-plugin/blob/c32075588ce3a6ea3bdb0afecc0bb609232112bf/src/it/update-file-header-test-mojo/pom.xml#L98 and I just checked that the test is correct. So I doubt the plugin is completely broken. Hard to say what is different in your case. Adding an integration test reproducing your situation under https://github.com/mojohaus/license-maven-plugin/tree/master/src/it  would help.

**Comment by chrisinmtown** [comment]
Thanks for the reply.  Working towards a test. I prepared a zip archive with two files, pom.xml and src/angular.json, trying to attach it here.  Just unpack and run "mvn license:update-file-header" to try. When I do this with mvn ver 3.5.2 it adds a header even tho the configuration is supposed to exclude *.json.

**Comment by ppalaga** [comment]
> I don't have all the infrastructure to update @Version@ automatically

You can run a single integration test by issuing 

    mvn clean verify -Dinvoker.test=update-file-header-test-mojo

from the root directory of the license-maven-plugin source tree.

> Just unpack and run "mvn license:update-file-header" to try.

Sorry, I prefer getting a pull request with an integration test. If this is really a bug, a fix will not be accepted without a test. By preparing a test, you help to deliver the fix.

**Comment by chrisinmtown** [comment]
Thanks for teaching me how to run tests! I issued that command in the root directory of the source tree, I thought it would run just one test but instead it ran many.  For example it ran test ISSUE-154 which failed. Is there an easy way to run a single test AND leave the output there instead of cleaning?

**Comment by ppalaga** [comment]
> I thought it would run just the one test

`-Dinvoker.test=update-file-header-test-mojo` should definitely make it run just the single test in `update-file-header-test-mojo` subdir of `src/it`.



> I am also a little puzzled that the pom.xml in the update-file-header-test-mojo directory omits the supposedly mandatory licenseName configuration property.

It is there https://github.com/mojohaus/license-maven-plugin/blob/c32075588ce3a6ea3bdb0afecc0bb609232112bf/src/it/update-file-header-test-mojo/pom.xml#L80

> When I run 'mvn license:update-file-header'

It may have to do with using and `<execution>` if you have your config under an `<execution>`, try to add  `<id>my-id</id>` under that `<execution>` and then run with `mvn license:update-file-header@my-id`. Only that way the config under the `<execution>` gets effective.

**Comment by chrisinmtown** [comment]
Re missing license name, I see it at plugin/executions/execution/configuration/licenseName but the l-m-p seems to require that property at plugin/configuration/licenseName when invoked only via mvn license:update-file-header (?).

So I think my stumbling block here is the maven lifecycle phase.  I guess during 'mvn verify' mvn goes thru all the usual phases including 'process-sources' so the configuration is used properly. But, again guessing, during 'mvn license:update-file-header' that phase is skipped, so the execution configuration does NOT apply, and my little exclusion is ignored. Does that sound right?

Basically  I'm trying to use the plugin as a one-and-done tool, and I've misconfigured it by using the wrong lifecycle phase.  I think your expectation is that the tool will be an integral part of the build lifecycle forever, updating headers as files appear or become outdated.

**Comment by ppalaga** [comment]
Which parts of config are effective with `mvn license:update-file-header` vs. `mvn license:update-file-header@my-execution` is given by Maven rather than license-maven-plugin. license-maven-plugin is not opinionated about whether the mojos are invoked standalone or bound to a phase. Do whatever suits you. Closing as you problem seems to be solved.

**Comment by chrisinmtown** [comment]
I'm ok with you closing this because I agree that no code change is needed to l-m-p. 

Before moving on, would you PLEASE explain how to specify the phase, or what magic to enter or type, so that files are properly excluded when I invoke via 'mvn license:update-file-header'?  That would be a big help from an expert in mojos (whatever that means).
